### ISSUE TITLE
Configure OidcProviderOptions using IConfiguration on Blazor Wasm

### ISSUE THREAD

**Issue by unaizorrilla** [dotnet/aspnetcore#20231]
## IConfiguration can't be used when configuring the host builder

There is no way to configure  AddOidcAuthentication using the IConfiguration of blazor wasm 

```csharp

            var builder = WebAssemblyHostBuilder.CreateDefault(args);
            builder.RootComponents.Add<App>("app");

            builder.Services.AddBaseAddressHttpClient();
            builder.Services.AddOidcAuthentication(options =>
            {
                options.ProviderOptions.Authority = "https://demo.identityserver.io";
                options.ProviderOptions.ClientId = "interactive.public";
                options.ProviderOptions.DefaultScopes.Add("api");
                options.ProviderOptions.ResponseType = "code";
            });
```

Ideally we need the IConfiguration information to put this configuration values on appsettings.json or appsettings.development.json etc but right now the only way is hardcoded this parameters on the Program.cs

### Further technical details
- Blazor Wasm 3.2.0-preview3.20168.3

**Comment by javiercn** [comment]
@unaizorrilla thanks for filing the issue.

We'll discuss it within the team and update it accordingly.

**Comment by dibble-james** [comment]
I've tried getting around this by building the configuration before configuring the OIDC provider like this:
```csharp
```
but that results in the

**Comment by rynowak** [comment]
One of the ideas here (that @davidfowl tried in FeatherHttp) is to make the configuration builder object an `IConfiguration` itself. That way the "current state" of configuration is readable, but the configuration itself is still mutable. 

This change would result in making a new public type in BlazorWASM's hosting later that implements this functionality.

**Comment by pranavkm** [comment]
I can look at fixing the issue @dibble-james ran in to where building the `IConfigurationBuilder` twice fails. Do we want to do anything specific for blazor-wasm to allow reading config without having to construct it? For instance, we could expose a `FooConfigurationBuilder : IConfigurationBuilder, IConfiguration` off of the host builder.

**Comment by dibble-james** [comment]
FYI if anyone wants a real dirty way around this while this filters through you can do something like this:
```csharp
public class ConfigurableRemoteAuthenticationService : RemoteAuthenticationService<RemoteAuthenticationState, OidcProviderOptions>
    {
        public ConfigurableRemoteAuthenticationService(
            IJSRuntime jsRuntime,
            IConfiguration configuration,
            IOptions<RemoteAuthenticationOptions<OidcProviderOptions>> options,
            NavigationManager navigation)
            : base(jsRuntime, BuildOptions(configuration, options.Value), navigation)
        {
        }

        private static OptionsWrapper<RemoteAuthenticationOptions<OidcProviderOptions>> BuildOptions(
            IConfiguration configuration,
            RemoteAuthenticationOptions<OidcProviderOptions> options)
        {
            options.ProviderOptions.Authority = configuration["Authority"];

            return new OptionsWrapper<RemoteAuthenticationOptions<OidcProviderOptions>>(options);
        }
    }
```
Then register it like so: `builder.Services.AddSingleton<AuthenticationStateProvider, ConfigurableRemoteAuthenticationService>();`

**Comment by MichaelSpranger** [comment]
The problem @dibble-james ran into is that a stream in general can only be read once, so when the framework is building the configuration again reading the stream fails. This is also the cause of the bug [Using MemoryStream on JsonConfigurationExtensions.AddJsonStream throws StreamNotReadable due to casting #1979](https://github.com/dotnet/extensions/issues/1979) which unfortunately was closed.
A possible workaround is to new up a ConfigurationBuilder and set it up with the same sources as the one of the WebAssemblyHostBuilder. This is far from ideal.
A better solution would be if configuration files are read into memory first in a way that supports reading multiple times, e.g. as a string. I do not know if there is already a method in the framework analogous to AddJsonStream to support this.

**Comment by MortenMeisler** [comment]
Alot of appsettings configuration happens in program.cs. Like authentication.

For DevOps I think this issue/feature is essential.

**Comment by MichaelSpranger** [comment]
I realized that I am able to access the configuration like so. I use `builder.Services.AddMsalAuthentication`, but I expect this to work the same with AddOidcAuthentication.
```cs
            builder.Services.AddMsalAuthentication(options =>
            {
                var configuration = builder.Services.BuildServiceProvider().GetRequiredService<IConfiguration>();
...
            });
```

**Comment by davidfowl** [comment]
That’s not right, calling BuidServiceProvider is wrong as it will result in another set of service instances

**Comment by MichaelSpranger** [comment]
@davidfowl Agree, this is not how it should be done, and lacking a proper solution this is what a lot of people will end up doing.

**Comment by yudielcurbelo** [comment]
> I realized that I am able to access the configuration like so. I use `builder.Services.AddMsalAuthentication`, but I expect this to work the same with AddOidcAuthentication.
> 
> ```cs
>             builder.Services.AddMsalAuthentication(options =>
>             {
>                 var configuration = builder.Services.BuildServiceProvider().GetRequiredService<IConfiguration>();
> ...
>             });
> ```

This is working for me in the meantime.

**Comment by captainsafia** [comment]
Closing this as it was resolved in #20647.

We'll be making changes to the Blazor WASM template to read from the new configuration property in the startup logic. This is tracked in #20646
